plutus-contract/src/Language/Plutus/Contract/StateMachine.hs

```diff
@@ -251,22 +268,23 @@ runGuardedStep ::
     => StateMachineClient state input              -- ^ The state machine
     -> input                                       -- ^ The input to apply to the state machine
     -> (UnbalancedTx -> state -> state -> Maybe a) -- ^ The guard to check before running the step
-    -> Contract schema e (Either a state)
-runGuardedStep smc input guard = mapError (review _SMContractError) $ do
-    StateMachineTransition{smtConstraints,smtOldState=State{stateData=os}, smtNewState=State{stateData=ns}, smtLookups} <- mkStep smc input
-    pk <- ownPubKey
-    let lookups = smtLookups { Constraints.slOwnPubkey = Just $ pubKeyHash pk }
-    utx <- either (throwing _ConstraintResolutionError) pure (Constraints.mkTx lookups smtConstraints)
-    case guard utx os ns of
-        Nothing -> do
-            submitTxConfirmed utx
-            pure $ Right ns
-        Just a  -> pure $ Left a
+    -> Contract schema e (GuardedStepResult state input a)
```

Comment:
I wonder if we should represent this as `Either a (TransitionResult state input)`, where a `TransitionResult` is either a successful transition or a failed one.

We could maybe reuse that later - there are a few places that seem to return something like that.

Reply:
Then again, if `TransitionResult` is basically just `Either InvalidTransition StateMachineTransition`, then I'm suggesting something like `Maybe (Either X Y)`, which is perhaps a bit obscure. I think it would be okay if at least the inner `Either` type was named... I'm not really sure tbh.

Reply:
Yes that would be better